added support for generation-based query and delete of blobs.

tsinha opened the pull request:
Small modifications to get_blob and delete_blob to handle generation ids when querying buckets that have versioning enabled.

tsinha commented:
I signed the CLA as Vital Labs, Inc.

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -222,14 +227,22 @@ def get_blob(self, blob_name, client=None):
+            path = blob.path
+            if generation is not None:
+                path = (blob.path + '?generation=' +
+                        quote(str(generation), safe=''))
```
We should move `generation` onto `blob` so that `blob.path` picks it up if need be.

tseaver replied:
That would require making it a ctor argment (which wouldn't be a bad idea, anyway).

dhermes replied:
Not here it wouldn't, since the `blob` is constructed two lines above, we could use a `generation` setter on the blob.

tsinha replied:
I did this originally.  But, after seeing how the code was coming together and thinking of the models a bit, I came to the conclusion that the bucket was a better place for this.  

Here is my reasoning; I'd love your feedback:

1) The blob object doesn't facilitate querying, rather it's a representation of what the storage system believes exists.  Allowing someone to define the generation of a blob, through a setter, seems to be counter to the purpose of the generation property on the blob.  That is, blob.generation is the canonical ID returned by the storage system for that blob.  Not something a user should be able to define.  

2) The blob.path is an imputed property and constructed from the only two key pieces of information needed to define a blob: bucket and name.  User-defined generation seems a bit superfluous, since it has no real meaning in the definition of a blob.

These two points led me to think of the generation as a query parameter better suited to the bucket API.

The problem with my approach now is that it disregards all other query parameters (ifGenerationNotMatch, ifMetagenerationNotMatch, etc.).  

I think the next step for this path of development would be a generator function, in the bucket, that uses user-specified query parameters and the blob.path to create the full API query URL.

Let me know your thoughts... if you would like the query param to be through a setter to be at the blob level, and corresponding blob.path construction to happen there, I can refactor.

dhermes replied:
I totally get why you'd go that route. A little of where how we envision these working together may help. We consider `Bucket.get_blob` and `Bucket.delete_blob` to be tiny do-nothing wrappers and put all verbs (i.e. API methods) relating to a noun (here the noun is `Blob`) on the relevant object.

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -214,6 +215,10 @@ def get_blob(self, blob_name, client=None):
+          >>> print bucket.get_blob(
+          ... '/path/to/versioned_blob.txt',
+          ... generation=generation_id)
+          <Blob: my-bucket, /path/to/versioned_blob.txt>
```
Thanks a lot for updating the docstring!

dhermes commented:
@tsinha You can ignore the AppVeyor failure for `test_upload_from_file_w_bound_client_multipart`, we are tracking it on #1434.

Updating `Bucket.delete_blob` and `Bucket.get_blob` is insufficient here, `Blob.delete` and `Blob.exists` and `Blob.reload` may also need updating (or you may get that for free when updating `Blob.path`)

@jgeewax What should be done about matching his signature of the CLA with his GitHub handle?

tsinha commented:
I'll look at the blob level methods and make sure everything works with the bucket level query, or refactor as you suggested if needed.  

I'll post an update shortly.

tsinha commented:
@dhermes I looked over and fixed blob.exists and blob.delete.

Still unsure about adding generation param directly into blob.path

It doesn't seem as clean as having the api_request create the appropriate paths via blob.path and query_params at the bucket level.

Otherwise, I think this PR should be ready to go.

dhermes commented:
@tsinha the generation is required to be in the path, since it uniquely defines the object if specified.

`Bucket.delete_blob` and `Bucket.get_blob` are meant to be "dumb" (i.e. they don't know how to do much) and the knowledge of how to make the request should live on `Blob`.

tsinha commented:
@dhermes Thanks for the feedback.
Can you help me brainstorm how to reconcile the way in which [api_requests are made](https://github.com/GoogleCloudPlatform/gcloud-python/blob/master/gcloud/connection.py#L182) with embedding the generation parameter at the blob level?

This confounds [my current WIP](https://github.com/VitalLabs/gcloud-python/blob/8d92cf55264ad8d2295103c1b320ee8cbe56e7c1/gcloud/storage/bucket.py#L492), which will enable the renaming of blobs while preserving versions, since the 'copyTo' API end-point requires another query param ('sourceGeneration') to copy the right version.

If we move generation into blob.path, I'd end up with something like the following:

https://www.googleapis.com/storage/v1/b/sourceBucket/o/sourceObject?generation=XXX/copyTo/b/destinationBucket/o/destinationObject?generation=YYY?sourceGeneration=XXX

But I may not be thinking it through completely... any advice you have would be helpful.

dhermes commented:
I think you've just thought about it more than I had :grinning: 

However, I'd still rather special case `Bucket.copy_blob` than special case every other place a generation might be used.

You could also leave `Blob.path` as-is and we could have another property that also contains identifying informating in query params, like `Blob.path_with_params` (or a name that is less hideous).

tsinha commented:
@dhermes I've added a new blob.path_with_params function to alleviate some of the checking at the bucket level: [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/blob.py#L137) and [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/bucket.py#L243).

I've also got a working version of the copy_blob and rename_blob functions with versioning support [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/bucket.py#L464) and [here](https://github.com/VitalLabs/gcloud-python/blob/master/gcloud/storage/bucket.py#L515).

I changed the API for copy_blob a little to fit into the existing rename_blob logic... that is, if versioned copy is enabled, copy_blob returns a tuple of two lists that contain the old and new blobs... rename_blobs takes the old_blobs list from this tuple and deletes them, and then returns the new_blobs list.

If versioning is not enabled, then copy_blobs's return parameters are unchanged from the existing API.

What do you think about this approach?  I can add testing around this if it looks reasonable.

dhermes commented:
@tsinha Sorry for the review delay. I'm traveling and won't have a chance to look at this until next Monday. (I feel bad, sorry for the negative experience.)

tsinha commented:
No worries... I look forward to your feedback next week. Safe travels.

dhermes commented on gcloud/storage/blob.py:
```diff
@@ -133,6 +134,22 @@ def path(self):
+        if self.generation is not None:
+            params = {'generation': self.generation}
+
+        return (self.path, params)
```
Parens are superflous

dhermes replied:
Also, I expected this to just be a string, though I'll hold off on comments until I see how it's used below.

dhermes commented on gcloud/storage/blob.py:
```diff
@@ -133,6 +134,22 @@ def path(self):
 
     @property
+    def path_with_params(self):
+        """Getter property for the URL path to this Blob, with version.
```
s/with version/with generation

dhermes commented on gcloud/storage/blob.py:
```diff
@@ -133,6 +134,22 @@ def path(self):
+        :rtype: tuple of ``path`` (a string) and ``params`` (a dictionary)
+        :returns: the URL path to this blob and a dictionary with the
+                 generation that can be used in query_params for
+                 connection.api_request
```
Indent 1 more space to align with `the` in `:returns: the`

dhermes commented on gcloud/storage/blob.py:
```diff
@@ -241,6 +258,9 @@ def exists(self, client=None):
             # minimize the returned payload.
             query_params = {'fields': 'name'}
+            if self.generation is not None:
+                query_params['generation'] = self.generation
```
Why not use `self.path_with_params` here? In my head, that was the point.

dhermes commented on gcloud/storage/blob.py:
```diff
@@ -266,7 +286,8 @@ def delete(self, client=None):
         """
-        return self.bucket.delete_blob(self.name, client=client)
+        return self.bucket.delete_blob(self.name, client=client,
+                                       generation=self.generation)
```
I had forgotten we give such little info to` Bucket.delete_blob` (i.e. it only has the blob's name, not the object itself).

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -103,7 +103,7 @@ def client(self):
         return self._client
 
-    def blob(self, blob_name, chunk_size=None):
+    def blob(self, blob_name, chunk_size=None, generation=None):
```
No need for this here. The `@setter` can handle it.

dhermes replied:
Reasoning: `generation` is a property of a blob, but there are too many properties to provide in the constructor

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -222,14 +232,20 @@ def get_blob(self, blob_name, client=None):
         """
         client = self._require_client(client)
-        blob = Blob(bucket=self, name=blob_name)
+        blob = Blob(bucket=self, name=blob_name, generation=generation)
```
RE: Comment above, you can just use the setter:

```python
blob = Blob(bucket=self, name=blob_name)
blob.generation = generation
```

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -392,10 +412,15 @@ def delete_blob(self, blob_name, client=None):
         blob_path = Blob.path_helper(self.path, blob_name)
+        query_params = {}
+        if generation is not None:
+            query_params = {'generation': generation}
```
Shouldn't this use the new method that includes the path params?

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -423,7 +448,13 @@ def delete_blobs(self, blobs, on_error=None, client=None):
-                self.delete_blob(blob_name, client=client)
+
+                generation = None
+                if hasattr(blob, 'generation'):
```
`generation` is a property so `hasattr` is not needed. You can bunch this in with the `not isinstance(blob_name, six.string_types)` above (since that represents a `Blob` object)

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -448,18 +479,38 @@ def copy_blob(self, blob, destination_bucket, new_name=None,
+        if versions:
+            return (new_blobs, old_blobs)
+        else:
+            return new_blobs[0]
```
Let's do this in another PR. Scope should not grow *during* review :grinning:

dhermes commented on gcloud/storage/bucket.py:
```diff
@@ -484,11 +535,19 @@ def rename_blob(self, blob, new_name, client=None):
+        if len(new_blobs) == 1:
+            return new_blobs[0]
+        else:
+            return new_blobs
```
Ditto here on tackling in another PR.

tsinha commented:
@dhermes I have not forgotten about this.  Tied up with some other stuff and will get back to this ASAP.

dhermes commented:
Thanks for the update

tillahoffmann commented:
@dhermes, @tsinha, thanks for looking into this. Do you have any updates on when this PR might get merged?

dhermes commented:
@tillahoffmann There have been lots of changes to `blob` since this PR was originally sent. Some code paths in the current release (`google-cloud-storage==1.2.0`) will "just work" for this feature.

Peeking at the [code][1], at least the `Blob.download_*` functions **will** incorporate the generation (but, as a hack, only if `media_link` isn't set on the `Blob` instance). As for [delete][2], it calls out to [`Blob.path_helper`][3] which **does not** incorporate the generation.

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/storage-1.2.0/storage/google/cloud/storage/blob.py#L382-L398
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/storage-1.2.0/storage/google/cloud/storage/bucket.py#L417
[3]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/storage-1.2.0/storage/google/cloud/storage/blob.py#L193

tillahoffmann commented:
@dhermes, thanks for the update. Do you have any recommendations on how to access md5 and crc32c hashes given a generation number from python?

dhermes commented:
Sure thing. Here is a somewhat icky hack (but it'll work for now).

For a file with multiple versions:

```
$ gsutil ls -a gs://some-bucket
gs://some-bucket/file.txt#1498874876972049
gs://some-bucket/file.txt#1498875135704497
```

If you `reload()` it like normal, you'll see the latest version:

```python
>>> from google.cloud import storage
>>>
>>> client = storage.Client()
>>> bucket = client.bucket('some-bucket')
>>> blob_normal = bucket.blob('file.txt')
>>> blob_normal.reload()
>>> blob_normal._properties
{u'bucket': u'some-bucket',
 u'contentType': u'text/plain',
 u'crc32c': u'JcxYOQ==',
 u'etag': u'CLH7moCB59QCEAE=',
 u'generation': u'1498875135704497',
 u'id': u'some-bucket/file.txt/1498875135704497',
 u'kind': u'storage#object',
 u'md5Hash': u'F/IyVDfyzIWJW12F0xO6+Q==',
 u'mediaLink': u'https://www.googleapis.com/download/storage/v1/b/some-bucket/o/file.txt?generation=1498875135704497&alt=media',
 u'metageneration': u'1',
 u'name': u'file.txt',
 u'selfLink': u'https://www.googleapis.com/storage/v1/b/some-bucket/o/file.txt',
 u'size': u'1672',
 u'storageClass': u'MULTI_REGIONAL',
 u'timeCreated': u'2017-07-01T02:12:15.663Z',
 u'timeStorageClassUpdated': u'2017-07-01T02:12:15.663Z',
 u'updated': u'2017-07-01T02:12:15.663Z'}
```

However, if you manually patch `reload` on the object, you can get the data you want:

```python
>>> from google.cloud.storage.blob import _quote
>>>
>>> GENERATION = '1498874876972049'
>>>
>>> def reload(self, client=None):
...     client = self._require_client(client)
...     query_params = {'projection': 'noAcl', 'generation': GENERATION}
...     api_response = client._connection.api_request(
...         method='GET', path=self.path, query_params=query_params,
...         _target_object=self)
...     self._set_properties(api_response)
...
>>>
>>> blob_patched = bucket.blob('file.txt')
>>> blob_patched.reload = reload.__get__(blob_patched)
>>> blob_patched._properties
{u'bucket': u'some-bucket',
 u'contentType': u'text/plain',
 u'crc32c': u'MO9tZQ==',
 u'etag': u'CJGY64SA59QCEAE=',
 u'generation': u'1498874876972049',
 u'id': u'some-bucket/file.txt/1498874876972049',
 u'kind': u'storage#object',
 u'md5Hash': u'zPc9wCyNcKGiYV3ijiB2iw==',
 u'mediaLink': u'https://www.googleapis.com/download/storage/v1/b/some-bucket/o/file.txt?generation=1498874876972049&alt=media',
 u'metageneration': u'1',
 u'name': u'file.txt',
 u'selfLink': u'https://www.googleapis.com/storage/v1/b/some-bucket/o/file.txt',
 u'size': u'1635',
 u'storageClass': u'MULTI_REGIONAL',
 u'timeCreated': u'2017-07-01T02:07:56.965Z',
 u'timeDeleted': u'2017-07-01T02:12:15.704Z',
 u'timeStorageClassUpdated': u'2017-07-01T02:07:56.965Z',
 u'updated': u'2017-07-01T02:07:56.965Z'}
```

lukesneeringer commented:
Hi @tillahoffmann,
Since the code has changed so much since your original PR, and it sounds like there is an existing path to do what you need now, I am going to close this. I am sorry that we sat on it for so long.

If there are any outstanding issues remaining, feel free to re-open (or open a new issue).
